lambda/matching.ml

```diff
@@ -2790,41 +2828,58 @@ let combine_constant value_kind loc arg cst partial ctx def
 let split_cases tag_lambda_list =
   let rec split_rec = function
     | [] -> ([], [])
-    | (cstr_tag, act) :: rem -> (
+    | ({cstr_tag; cstr_repr; cstr_constant}, act) :: rem -> (
         let consts, nonconsts = split_rec rem in
-        match cstr_tag with
-        | Cstr_constant n -> ((n, act) :: consts, nonconsts)
-        | Cstr_block n -> (consts, (n, act) :: nonconsts)
-        | Cstr_unboxed -> (consts, (0, act) :: nonconsts)
-        | Cstr_extension _ -> assert false
+        match cstr_tag, cstr_repr with
+        | Ordinary _, Variant_unboxed _ -> (consts, (0, act) :: nonconsts)
+        | Ordinary {runtime_tag}, Variant_boxed _ when cstr_constant ->
+          ((runtime_tag, act) :: consts, nonconsts)
+        | Ordinary {runtime_tag}, Variant_boxed _ ->
+          (consts, (runtime_tag, act) :: nonconsts)
+        | _, Variant_extensible -> assert false
+        | Extension _, _ -> assert false
       )
   in
   let const, nonconst = split_rec tag_lambda_list in
   (sort_int_lambda_list const, sort_int_lambda_list nonconst)
 
+(* The bool tracks whether the constructor is constant, because we don't have a
+   constructor_description available for polymorphic variants *)
+let split_variant_cases (tag_lambda_list : ((int * bool) * lambda) list) =
+  let rec split_rec = function
+    | [] -> ([], [])
+    | (tag, act) :: rem -> (
+        let consts, nonconsts = split_rec rem in
+        match tag with
+        | (n, true) -> ((n, act) :: consts, nonconsts)
+        | (n, false) -> (consts, (n, act) :: nonconsts)
+      )
+  in
+  let const, nonconst = split_rec tag_lambda_list in
+  (sort_int_lambda_list const, sort_int_lambda_list nonconst)
+
+
 let split_extension_cases tag_lambda_list =
   let rec split_rec = function
     | [] -> ([], [])
-    | (cstr_tag, act) :: rem -> (
+    | ({cstr_constant; cstr_tag}, act) :: rem -> (
```

Comment:
This also looks like a `partition_map`.

Reply:
I've changed these.  I hesitated to do so for two reasons:
1) Minimizing the diff in this patch to the relevant bits (here, `split_extension_cases` is a pre-existing function with this logic, and the other is meant to mirror it.  But clearly that ship has sailed in this PR.
2) Now we're allocating extra "Either"s.

But in the end I agree it's an improvement.